runtime/src/main/java/org/corfudb/runtime/view/stream/BackpointerStreamView.java

```diff
@@ -0,0 +1,367 @@
+                log.trace("Overwrite occurred at {}", tokenResponse);
+                // We got overwritten, so we call the deacquisition callback
+                // to inform the client we didn't get the address.
+                if (deacquisitionCallback != null) {
+                    if (!deacquisitionCallback.apply(tokenResponse)) {
+                        log.debug("Deacquisition requested abort");
+                        return -1L;
+                    }
+                }
+                // Request a new token, informing the sequencer we were
+                // overwritten.
+                tokenResponse = runtime.getSequencerView()
+                        .nextToken(Collections.singleton(streamID),
+                             1, true, false);
+            }
+        }
+    }
+
+    /** {@inheritDoc}
+     *
+     * The backpointer version of remaining() calls nextUpTo() multiple times,
+     * as it uses the default implementation in IStreamView. While this may
+     * appear to be non-optimized, these reads will most likely hit in the
+     * address space cache since the entries were read in order to resolve the
+     * backpointers.
+     * */
+    public synchronized LogData nextUpTo(long maxGlobal) {
+        // Pop the context if it has changed.
+        if (getCurrentContext().logPointer >=
+                getCurrentContext().maxAddress) {
+            StreamContext last = streamContexts.pollFirst();
+            log.trace("Completed context {}@{}, removing.",
+                    last.contextID, last.maxAddress);
+        }
+
+        // If we have no entries to read, fill the read queue.
+        if (getCurrentContext().readQueue.isEmpty()) {
+            fillReadQueue(maxGlobal);
+            // If we still have no entries to read, there's nothing in
+            // the stream (we linearized the read during fillReadQueue()).
+            if (getCurrentContext().readQueue.isEmpty()) {
+                return null;
+            }
+        }
+
+        // If the lowest value in the read queue is greater than
+        // max global, there's nothing for us to do.
+        if (getCurrentContext().readQueue.first() > maxGlobal) {
+            return null;
+        }
+
+        // Otherwise we remove entries one at a time from the read queue,
+        // reading each entry in order until one of the entries contains
+        // data -or- we exceed maxGlobal -or- there are no more entries. If
+        // we exceed maxGlobal or run out of entries, we return null, else
+        // we return the data.;
+
+        while ( getCurrentContext().readQueue.size() > 0 &&
+                getCurrentContext().readQueue.first() <= maxGlobal) {
+            final long thisRead = getCurrentContext().readQueue.pollFirst();
+            getCurrentContext().logPointer = thisRead;
+            LogData ld = runtime.getAddressSpaceView().read(thisRead);
+            if (ld.containsStream(getCurrentContext().contextID)) {
```

Comment:
since this is the common and desired case, it is not nice to have the entire body inside an if statement.

make the inverse if test and 'continue'. Then proceed with the normal case.